libcasr/src/js.rs

```diff
@@ -0,0 +1,390 @@
+//! JS module implements `ParseStacktrace` and `Exception` traits for JS reports.
+use crate::error::{Error, Result};
+use crate::exception::Exception;
+use crate::execution_class::ExecutionClass;
+use crate::stacktrace::{ParseStacktrace, Stacktrace, StacktraceEntry};
+
+use regex::Regex;
+
+/// Structure provides an interface for parsing JS exception message.
+pub struct JsException;
+
+impl Exception for JsException {
+    fn parse_exception(stderr: &str) -> Option<ExecutionClass> {
+        let rexception = Regex::new(r"(?m)^.*?(\S*Error):(?:\s+(.*))?$").unwrap();
+        let Some(captures) = rexception.captures(stderr) else {
+            return None;
+        };
+        let error_type = captures.get(1).unwrap().as_str();
+        let message = if let Some(message) = captures.get(2) {
+            message.as_str()
+        } else {
+            ""
+        };
+        Some(ExecutionClass::new((
+            "NOT_EXPLOITABLE",
+            error_type,
+            message,
+            "",
+        )))
+    }
+}
+
+/// Structure provides an interface for processing the stack trace.
+pub struct JsStacktrace;
+
+impl ParseStacktrace for JsStacktrace {
+    fn extract_stacktrace(stream: &str) -> Result<Vec<String>> {
+        // Get stack trace from JS report.
+        let re = Regex::new(r"(?m)^(?:(?:.*Error:)(?:.|\n)*?|Thrown at:\n*?)((?:\n\s*at \S+.*)+)")
+            .unwrap();
+        let Some(cap) = re.captures(stream) else {
+            return Err(Error::Casr(
+                "Couldn't find traceback in JS report".to_string(),
+            ));
+        };
+        let mut stacktrace = cap
+            .get(1)
+            .unwrap()
+            .as_str()
+            .split('\n')
+            .filter(|l| !l.is_empty())
+            .map(|l| l.trim().to_string())
+            .collect::<Vec<String>>();
+
+        // Filter out multiple spaces
+        let re = Regex::new(r" +").unwrap();
+        stacktrace = stacktrace
+            .iter()
+            .map(|e| re.replace_all(e, " ").to_string())
+            .collect();
+
+        Ok(stacktrace)
+    }
+
+    fn parse_stacktrace_entry(entry: &str) -> Result<StacktraceEntry> {
+        let mut stentry = StacktraceEntry::default();
+        let re_full = Regex::new(r"^\s*at\s+(.+?)(?:\s+(\[as.*?\])?\s*)\((.*)\)$").unwrap();
+        let re_without_pars = Regex::new(r"^\s*at\s+(.+?)(?:(?:\s+(\[as.*?\]))?\s*)$").unwrap();
+
+        /// Parse substring of `entry` related to location in some file
+        ///
+        /// # Arguments
+        ///
+        /// * `entry` - initial string for stacktrace entry
+        ///
+        /// * `stentry` - `StacktraceEntry` to save parsed location into
+        ///
+        /// * `loc` - substring to extract location information from
+        fn parse_location(entry: &str, stentry: &mut StacktraceEntry, loc: &str) -> Result<()> {
```

Comment:
Why we need  output mutable parameter? Maybe just return as `Result<DebugInfo>`?